### DISCUSSION TITLE
Watermark for images

### DISCUSSION THREAD

**Discussion by workatease** [directus/directus#4208] [Feature Requests]
It would be a great if add a option to add watermark to uploaded image
This can be managed via Storage asset presets

**Comment by rijkvanzanten** [comment]
`sharp` (the image manipulation library we use) has a `composite` function to layer an image on top of the base image https://sharp.pixelplumbing.com/api-composite

I'm wondering if the inclusion of a watermark would be per asset "key", or based on a flag on directus_files, or.. 🤔 

Would you ever have the need for multiple watermarks, where you can choose based on the used key? What about permissions, should there be a way to "bypass" the watermark?

**Reply by workatease** [reply]
Yes we might have use cases key based and it would be good idea to support multiple watermark. 
bypassing watermark also make sense.

**Reply by workatease** [reply]
the usecase I see is for an app like Shutterstock when public images would be with watermark and if a customer has purchased the images the asset will be available based on the request of size and quality but without watermark

**Reply by rijkvanzanten** [reply]
yeah that's exactly what i was thinking too. In that case, we need to be able to dynamically enable/disable the watermark based on the current role, which means we might want to base it on some field in directus_files 🤔 

Also not sure how 80/20 it is at this point, we might want to consider adding a "on-asset-generation" hook or something like that so you could do this on the fly in a hook (@benhaynes

**Reply by benhaynes** [reply]
I think there are a _lot_ of cool things we can do with the asset generation endpoint, and this is one of them. Being able to set watermarks is nice, but doing it based on permissions is ideal. It's probably best to look into all these enhancements at one time... once our core platform is nice and stable.

**Comment by balavivek143** [comment]
Any idea when this feature will be materialized?  It seems like a good one and we are looking similar to this.

**Reply by rijkvanzanten** [reply]
[**Unfortunately, we can not provide estimates on features or fixes**](https://docs.directus.io/getting-started/support.html#can-you-give-an-eta-for-this-feature-fix). As of now, the only way to get an accurate timeline or release date is to [sponsor](https://docs.directus.io/getting-started/support.html#sponsored-work) this task. We recognize that this may be an important feature/fix, but we are a small open-source organization with a lot to triage and complete. If you can't sponsor this ticket, then the next best thing is to increase its priority by giving it a 👍 . Or contribute a pull-request... after all, this is OSS!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by vs-johnny-huang** [comment]
Share a workaround approach to add watermark inspired by the Sharp `composite` function @rijkvanzanten mentioned above, the solution isn't perfect but works as intended.

Steps
1. Prepare a watermark-only image PNG file and put it in a space (In my case, I pack it into Directus docker image together) that the Directus server be able to access.

2. Go to the _project settings page_ and add a new _transformation preset_ with _additional transformation_ as below (remember to replace the `file-system-absolute-path-to-watermark-file` with yours.)

update:  relative path works as well

```
[
    [
        "composite",
        [
            {
                "input": <file-system-absolute-path-to-watermark-file>,
                "tile": true
            }
        ]
    ]
]
```

4. After applying the _project settings_(you have to apply _project settings_ after you add a new preset), you can get a new image with watermark via the endpoint `<server-path>/assets/<file-id>/<filename>?key=<preset-key>` or `<server-path>/assets/<file-id>?key=<preset-key>`.

Hope that helps anyone who finds the discussion. It works so far so good at the moment on Directus 10.6.3.
